spec/models/article_spec.rb

```diff
@@ -1421,6 +1421,47 @@ def foo():
     end
   end
 
+  # rubocop:disable RSpec/AnyInstance
+  describe "#detect_language" do
```

Comment:
This is probably mostly _fine_ but, at the same time, having spent a _lot_ of the last week thinking about flaky tests specifically and test "discipline" more broadly, there are a few things we could do here to improve this:

* If we introduce a service object as an "adapter" around the `CLD3` interface, the `article` specs can focus on "article things" (like persistence, say)
* With an adapter class, we can write tests specific to testing language (maybe even probability threshold) that don't require hitting any database at all, and maybe don't ever need to build objects more complicated than `String`.
* We don't need to use `update_column` — we'll hit the DB three times for each of the new specs. Instead, if we can need to test the callback, we can use the `article` factory to instantiate a new article with any necessary pre-conditions — we can even use `build` for any tests that aren't specifically testing `update`-related behaviors.

An adapter can be something like:

```ruby
class LanguageDetector
  attr_reader :text
  PROBABILITY_THRESHOLD = 0.5

  def self.call(...)
    new(...).call
  end

  def initialize(text)
    @text = text
  end

  def call(identifier: CLD3::NNetLanguageIdentifier.new(0, 1000))
    language_outcome = identifier.find_language(text)
    return unless language_outcome.probability > PROBABILITY_THRESHOLD && language_outcome.reliable?
    language_outcome.language
  end
end
```

Then `Article#detect_language` can be something like:

```ruby
  def detect_language
    return unless title_changed? || body_markdown_changed?

    self.language = LanguageDetector.call("#{title}. #{body_text}")
  end
```

(The above is not pseudo code, it actually works. I couldn't figure out how to add as a `suggestion` on GitHub, but you can drop it in and it'll work.)

Then we can move the tests that are exploring `CLD3`-specific features (like probability threshold) to specs for the detector specifically and the `article` specs will run faster and be easier to follow. And _none_ of the specs would need to use `any_instance`, which we _really_ shouldn't use.

Reply:
Thanks for this @jaw6 — definitely no need to settle on "fine" when we're introducing a new thing and have a chance to improve the system overall where we can, so I'll rewrite these parts.